projects/lights_control/src/lights_blinker.c

```diff
@@ -0,0 +1,67 @@
+#include <stddef.h>
+
+#include "event_queue.h"
+#include "soft_timer.h"
+#include "status.h"
+
+#include "lights_blinker.h"
+
+static void prv_timer_callback(SoftTimerID timer_id, void *context) {
+  LightsBlinker *blinker = (LightsBlinker *)context;
+  blinker->state = !blinker->state;
+  event_raise(blinker->event_id, (uint16_t)blinker->state);
+  soft_timer_start_millis(blinker->duration_ms, prv_timer_callback, (void *)blinker,
+                          &blinker->timer_id);
+}
+
+// Blinker's timer id needs to be initialized to an invalid timer if it's not being used otherwise
+// we may have a collision.
+StatusCode lights_blinker_init(LightsBlinker *blinker) {
+  blinker->timer_id = SOFT_TIMER_INVALID_TIMER;
+  return STATUS_CODE_OK;
+}
+
+StatusCode lights_blinker_on(LightsBlinker *blinker, LightsBlinkerDuration duration_ms,
+                             EventID id) {
+  if (lights_blinker_inuse(blinker)) {
+    // The passed-in blinker may have active timers.
+    return status_msg(STATUS_CODE_INVALID_ARGS,
+                      "Can't call lights_blinker_on on an already ON blinker.");
+  }
+  blinker->duration_ms = duration_ms;
+  blinker->state = LIGHTS_BLINKER_STATE_ON;
+  blinker->event_id = id;
+  status_ok_or_return(event_raise(blinker->event_id, (uint16_t)blinker->state));
+  return soft_timer_start_millis(blinker->duration_ms, prv_timer_callback, (void *)blinker,
+                                 &blinker->timer_id);
+}
+
+StatusCode lights_blinker_off(LightsBlinker *blinker) {
+  if (blinker->timer_id == SOFT_TIMER_INVALID_TIMER) {
+    return status_msg(STATUS_CODE_INTERNAL_ERROR,
+                      "An already off blinker cannot be turned off again.");
+  }
+  blinker->state = LIGHTS_BLINKER_STATE_OFF;
+  status_ok_or_return(event_raise(blinker->event_id, (uint16_t)blinker->state));
```

Comment:
You may want to be careful with the order in which events happen and if things should be in a critical section. Imagine this:

```
blinker state = off
raise event
> soft timer interrupt
> blinker state = on
> raise event
> soft timer start (potentially with new ID?)
soft timer cancel (potentially with old ID?)
```

Same thing goes for all your functions.

Reply:
How can I put things in a critical section? Do I have control over making sure this completes before interrupts?

Reply:
Prefer to reorder things rather than use critical sections. Critical sections block all interrupts from running, not just your specific timer. Look at `critical_section.h` for critical sections.